doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst

```diff
@@ -0,0 +1,93 @@
+.. _ipc_service_backend_icmsg_with_buf:
+
+ICMsg with dynamically allocated buffers backend
+################################################
+
+This backend is built on top of the :ref:`ipc_service_backend_icmsg`.
+Data transferred over this backend travels in dynamically allocated
+buffers on shared memory. The ICMsg is used to send just references to
+those buffers. It also supports multiple endpoints.
+
+This architecture allows for overcoming some common problems
+with other backends (mostly related to multithread access and zero-copy).
+That is why this backend provides an alternative that has no known significant
+limitations like the others.
+
+Overview
+========
+
+The shared memory is divided into two parts. One is reserved for the ICMsg and
+the other contains equal-sized blocks. The number of blocks is configured in
+the devicetree.
+
+The data sending process is following:
+
+* The sender allocates one or more blocks. If there are not enough sequential
+  blocks, it waits using the timeout provided in the parameter which also
+  includes K_FOREVER and K_NO_WAIT.
+* The allocated blocks are filled with data. For the zero-copy case, this is
+  done by the caller, otherwise, it is copied automatically.
+  During this time other threads are not blocked in any way as long as there are
+  enough free blocks for them. They can allocate, send data and receive data.
+* A message containing block index is sent over ICMsg to the receiver.
+  The size of the ICMsg queue is big enough to hold messages for all blocks, so
+  it will never overflow.
```

Comment:
```suggestion
* A message containing the block index is sent over ICMsg to the receiver.
  The size of the ICMsg queue is large enough to hold messages for all blocks, so
  it will never overflow.
```